internal/jobtracker/jobtracker.go

```diff
@@ -0,0 +1,151 @@
+// and retrieve a stop signal for a specific job.
+type JobTrackerStore interface {
+	AddJob(ctx context.Context, jobType string) (uuid.UUID, error)
+	SetJobRunning(ctx context.Context, jobId uuid.UUID) error
+	SetJobSuccessful(ctx context.Context, jobId uuid.UUID) error
+	SetJobFailed(ctx context.Context, jobId uuid.UUID, jobErr error) error
+	GetJobStopSignal(ctx context.Context, jobId uuid.UUID) (stopSignal bool, err error)
+}
+
+// Tracker manages job tracking operations using a provided JobTrackerStore.
+// It periodically performs tasks based on the specified stopInterval duration.
+type Tracker struct {
+	store        JobTrackerStore
+	stopInterval time.Duration
+}
+
+// NewJobTracker creates and returns a new Tracker instance using the provided JobTrackerStore and stopInterval.
+// It returns an error if the store is nil or if stopInterval is not greater than zero.
+func NewJobTracker(store JobTrackerStore, stopInterval time.Duration) (*Tracker, error) {
+	tracker := &Tracker{}
+	if store == nil {
+		return tracker, errors.New("store cannot be nil")
+	}
+	if stopInterval <= 0 {
+		return tracker, errors.New("stopInterval must be greater than zero")
+	}
+	tracker.stopInterval = stopInterval
+	tracker.store = store
+
+	return tracker, nil
+}
+
+// Run runs a new job and reurns the job ID.
+func (j *Tracker) Run(ctx context.Context, jobType string, job func(ctx context.Context) error, deadline time.Duration) (uuid.UUID, error) {
+	jobId, err := j.store.AddJob(ctx, jobType)
+	if err != nil {
+		return jobId, err
+	}
+
+	go j.manageJob(ctx, jobId, deadline, job)
+
+	return jobId, nil
+}
+
+// manageJob runs a job with a given context, job ID, polling interval, and deadline.
+// It manages the job's lifecycle, including setting its status in the store, handling retries on store operations,
+// and responding to stop signals or context cancellations. The job is run in a separate goroutine, and its status
+// is updated as running, successful, or failed based on its result or context expiration.
+// If a stop signal is received or the context is canceled or times out, the job is marked as failed.
+// Store operations are retried up to 5 times with a 30-second delay between attempts in case of transient errors.
+func (j *Tracker) manageJob(
+	ctx context.Context,
+	id uuid.UUID,
+	deadline time.Duration,
+	job func(ctx context.Context) error,
+) {
+	jobCtx, cancelJob := context.WithTimeout(ctx, deadline)
+	defer cancelJob()
+
+	jobErrCh := make(chan error)
+
+	go j.runJob(ctx, jobCtx, id, jobErrCh, job)
+	j.monitorJob(ctx, jobCtx, id, jobErrCh, cancelJob)
+}
+
+func (j *Tracker) runJob(ctx, jobCtx context.Context, id uuid.UUID, jobErrCh chan error, job func(context.Context) error) {
+	if err := j.store.SetJobRunning(ctx, id); err != nil {
+		jobErrCh <- fmt.Errorf("failed to set job running, job not starting: %w", err)
+		return
+	}
+	err := job(jobCtx)
+	jobErrCh <- err
```

Comment:
i think you should also cancel the `jobCtx` once the job returns, so that `monitorJob` can stop, too.

Reply:
or have an alternative way to say "this job is done. stop monitoring"

Reply:
also.. maybe set the job status to "done" once it's done

Reply:
If I cancel it, it'll could hit the cancelled condition and be returned as failed. We'd be hoping that:
```go
	jobErrCh <- err
        cancelJobCtx() // always run afterwards?
```

Reply:
then
`jobErrCh <- job(jobCtx)`

Reply:
what i mean is don't do 
```
err := job(jobCtx)
jobErrCh <- err
```
if you can do 
```
jobErrCh <- job(jobCtx)
```
in a single line..